web/src/main/java/org/openmrs/web/filter/GZIPRequestStream.java

```diff
@@ -0,0 +1,107 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.web.filter;
+
+import org.springframework.http.HttpEntity;
+
+import javax.servlet.ServletInputStream;
+import javax.servlet.ServletOutputStream;
+import javax.servlet.http.HttpServletRequest;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.zip.GZIPInputStream;
+
+/**
+ * Wraps Response Stream for GZipFilter
+ * 
+ * @author Matt Raible
+ * @version $Revision: 1.3 $ $Date: 2004/05/16 02:17:00 $
+ */
+public class GZIPRequestStream extends ServletInputStream {
+	
+	// abstraction of the output stream used for compression
+	protected OutputStream bufferedOutput = null;
+	
+	// state keeping variable for if close() has been called
+	protected boolean closed = false;
+	
+	// reference to original response
+	protected HttpServletRequest request = null;
+	
+	// reference to the output stream to the client's browser
+	protected ServletOutputStream output = null;
+	
+	protected GZIPInputStream zipInput = null;
+	
+	// reference to the input stream
+	protected ServletInputStream input = null;
+	
+	// default size of the in-memory buffer
+	private int bufferSize = 50000;
+	
+	//    /**
+	//     * Gets the input stream from a response entity. If the entity is gzipped then this will get a stream over the uncompressed data.
+	//     * @param entity the entity whose content should be read
+	//     * @return the input stream to read from
+	//     * @throws IOException
+	//     */
+	//    public static ServletInputStream getUngzippedContent(ServletInputStream responseStream) throws IOException {
+	//        if (responseStream == null) {
+	//            return responseStream;
+	//        }
+	//        if(contentEncoding.contains("gzip"))   responseStream=new GZIPInputStream(responseStream);
+	//        return responseStream;
+	//    }
+	
+	public GZIPRequestStream(HttpServletRequest request) throws IOException {
+		super();
+		closed = false;
+		this.request = request;
+		this.input = request.getInputStream();
+		this.zipInput = new GZIPInputStream(input);
+		
+		bufferedOutput = new ByteArrayOutputStream();
+	}
+	
+	public int read() throws IOException {
+		return input.read();
```

Comment:
reading from zipInput throws IOException in bufReader.readLine() in ZipFilterTest